### DISCUSSION TITLE
Timeout Error on Sending ACK back to Sender (Large Messages)

### DISCUSSION THREAD

**Discussion by muhammadali-ts** [nextgenhealthcare/connect#6375] [Q&A]
**Problem Description:**

I'm experiencing a timeout issue when sending an ACK back to the sender after receiving large HL7 messages (13-14 MB) that contain base64-encoded files (docx, PDFs, images). Even after configuring the source connector to auto-generate (before processing), the ACK still takes too long to be sent, resulting in a timeout on the sender's side.

**Source Connector Configuration:**

- Connector Type: TCP Listener
- Local Address: All Interfaces
- Source Queue: Off (Respond after processing)
- Response: Auto-generate (before processing)
- Process Batch: No
- Max Processing Threads: 1
- TCP Listener Settings:
  - Transmission Mode: MLLP
  - Mode: Server
  - Keep Connection Open: Yes

**Environment:**

- Mirth Connect Version: 4.5.0
- Java Version: 22
- Database for mirth Logging: MySQL
- Memory allocation:
  - Xmx: 4096M

**Assumptions:**

I believe the delay is caused by:
- Message Storage: Mirth Connect taking too long to store the large message in the database.
- Message Parsing: The size of the message affecting how long Mirth takes to parse it into components.

(Please correct me if I am wrong or if I am skipping any step)

**Constraints:**

Unfortunately, any change on sender side is not encouraged, such as:

- Increasing the sender's timeout beyond 15 seconds or set it to 0.
- Splitting the message into smaller parts to send.

**Question:** 

- Is there a way to optimize the ACK response time when dealing with large messages?

Any suggestions or solutions would be greatly appreciated! Thanks in advance.

**Comment by kirbykn2** [comment]
Have you tried turning the source queue on?

On Tue, Dec 10, 2024 at 10:05 AM Muhammad Ali ***@***.***>
wrote:

> *Problem Description:*
>
> I'm experiencing a timeout issue when sending an ACK back to the sender
> after receiving large HL7 messages (13-14 MB) that contain base64-encoded
> files (docx, PDFs, images). Even after configuring the source connector to
> auto-generate (before processing), the ACK still takes too long to be sent,
> resulting in a timeout on the sender's side.
>
> *Source Connector Configuration:*
>
>    - Connector Type: TCP Listener
>    - Local Address: All Interfaces
>    - Source Queue: Off (Respond after processing)
>    - Response: Auto-generate (before processing)
>    - Process Batch: No
>    - Max Processing Threads: 1
>    - TCP Listener Settings:
>       - Transmission Mode: MLLP
>       - Mode: Server
>       - Keep Connection Open: Yes
>
> *Environment:*
>
>    - Mirth Connect Version: 4.5.0
>    - Java Version: 22
>    - Database for mirth Logging: MySQL
>    - Memory allocation:
>       - Xmx: 4096M
>
> *Assumptions:*
>
> I believe the delay is caused by:
>
>    - Message Storage: Mirth Connect taking too long to store the large
>    message in the database.
>    - Message Parsing: The size of the message affecting how long Mirth
>    takes to parse it into components.
>
> (Please correct me if I am wrong or if I am skipping any step)
>
> *Constraints:*
>
> Unfortunately, any change on sender side is not encouraged, such as:
>
>    - Increasing the sender's timeout beyond 15 seconds or set it to 0.
>    - Splitting the message into smaller parts to send.
>
> *Question:*
>
>    - Is there a way to optimize the ACK response time when dealing with
>    large messages?
>
> Any suggestions or solutions would be greatly appreciated! Thanks in
> advance.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nextgenhealthcare/connect/discussions/6375>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APRXWD5YAOHIE7TIA7DBFEL2E37MHAVCNFSM6AAAAABTLMBOACVHI2DSMVQWIX3LMV43ERDJONRXK43TNFXW4OZXGY2TAOBXGE>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Best,

Kirby Knight
| 231.735.4650 | ***@***.***

**Reply by muhammadali-ts** [reply]
hey @kirbykn2 ,
Thanks for your reply

I have tried it and yes it works 
Now my next step is to validate message on certain rules 
and send ACK accordingly

is there any method for that you could guide me?
I have tried putting validation in source transformer but unfortunately it results in timeout on sender side

**Comment by kirbykn2** [comment]
If the other side is unwilling to adjust the timeout, you need to make a
choice.  Acknowledge before processing (validating) the message, or
validate the message and then respond, possible causing a timeout.

On Wed, Dec 11, 2024 at 6:47 AM Muhammad Ali ***@***.***>
wrote:

> hey @kirbykn2 <https://github.com/kirbykn2> ,
> Thanks for your reply
>
> I have tried it and yes it works
> Now my next step is to validate message on certain rules
> and send ACK accordingly
>
> is there any method for that you could guide me?
> I have tried putting validation in source transformer but unfortunately it
> results in timeout on sender side
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nextgenhealthcare/connect/discussions/6375#discussioncomment-11532138>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APRXWD4XC5UDC7TVFDA6T232FAQ4FAVCNFSM6AAAAABTLMBOACVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCNJTGIYTGOA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 
Best,

Kirby Knight
| 231.735.4650 | ***@***.***

**Reply by pacmano1** [reply]
Why is the other side unwilling to adjust their time out?    

Also, I suspect the actual transmission is what takes time here, not the ACK. You can check those in the message viewer.  How are you receiving the message from a topology perspective?  Does a VPN or other network need tuning?  

source -> ------ 10 seconds for 10MB (or whatever) ------> mirth ACK (split second normally).
